### PR TITLE
Generalize trace validation of byte_extract_exprt to include all endiannesses in JBMC

### PR DESCRIPTION
**Pull request by jeannielynnmoulton**
- [x] Each commit message has a non-empty body, explaining why the change was made.
- [x] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [x] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [x] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [x] My commit message includes data points confirming performance improvements (if claimed).
- [x] My PR is restricted to a single feature or bugfix.
- [x] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

### CODE REVIEW

File: unit/util/expr_cast/expr_cast.cpp [hunk 1]
```diff
@@ -141,6 +142,67 @@ SCENARIO("expr_dynamic_cast",
...
+    byte.op() = symbol_exprt(typet());
+    byte.offset() = constant_exprt("0", typet());
+    THEN("can_expr_expr<byte_extract_expr> returns true")
+    {
+      REQUIRE(can_cast_expr<byte_extract_exprt>(byte));
+    }
+  }
+  GIVEN("A byte extract expression with big endianness")
+  {
+    auto byte = byte_extract_exprt(ID_byte_extract_little_endian);
```
Comment (owen-mc-diffblue) [inline unit/util/expr_cast/expr_cast.cpp]: You mean `big_endian`

**Review by owen-mc-diffblue — APPROVED** [review]

File: src/util/byte_operators.h [hunk 2]
```diff
@@ -57,6 +57,13 @@ class byte_extract_exprt:public binary_exprt
   const exprt &offset() const { return op1(); }
 };
 
+template <>
+inline bool can_cast_expr<byte_extract_exprt>(const exprt &base)
+{
+  return base.id() == ID_byte_extract_little_endian ||
+         base.id() == ID_byte_extract_big_endian;
+}
+
```
Comment (thomasspriggs) [inline src/util/byte_operators.h:66]: :pick: I guess we should probably implement `validate_expr` at the same time as `can_cast_expr`, for consistency with other classes which inherit from `binary_exprt`.
```
inline void validate_expr(const byte_extract_exprt &value)
{
  validate_operands(value, 2, "byte extract must have two operands");
}
```

**Review by thomasspriggs — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: b46d3b775122222fa742c03e4c5208203db2ffbb).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117287261

**Review by LAJW — APPROVED** [review]

**Review by thk123 — APPROVED** [review]

**Review by allredj — COMMENTED** [review]
:heavy_check_mark:
Passed Diffblue compatibility checks (cbmc commit: 86b4fb3f2fd39afae9aac091fb828cc000398d58).
Build URL: https://travis-ci.com/diffblue/test-gen/builds/117298045

File: jbmc/src/java_bytecode/java_trace_validation.cpp [hunk 3]
```diff
@@ -251,23 +251,23 @@ static void check_rhs_assumptions(
     }
   }
   // check byte extract rhs structure
-  else if(rhs.id() == ID_byte_extract_little_endian)
+  else if(const auto byte = expr_try_dynamic_cast<byte_extract_exprt>(rhs))
```
Comment (smowton) [inline jbmc/src/java_bytecode/java_trace_validation.cpp:254]: byte_extract_expr

**Review by smowton — APPROVED** [review]
One nitpick otherwise lgtm

**Pull request merged**